internal/hcp/internal/controllers/telemetrystate/controller_test.go

```diff
@@ -0,0 +1,137 @@
+		WithTenancies(suite.tenancies...).
+		Run(suite.T())
+
+	suite.rt = controller.Runtime{
+		Client: client,
+		Logger: testutil.Logger(suite.T()),
+	}
+	suite.client = rtest.NewClient(client)
+}
+
+func TestTelemetryStateController(t *testing.T) {
+	suite.Run(t, new(controllerSuite))
+}
+
+func (suite *controllerSuite) deleteResourceFunc(id *pbresource.ID) func() {
+	return func() {
+		suite.client.MustDelete(suite.T(), id)
+	}
+}
+
+func (suite *controllerSuite) TestController_Ok() {
+	// Run the controller manager
+	mgr := controller.NewManager(suite.client, suite.rt.Logger)
+	mockClient, mockClientFn := mockHcpClientFn(suite.T())
+	mockClient.EXPECT().FetchTelemetryConfig(mock.Anything).Return(&hcpclient.TelemetryConfig{
+		MetricsConfig: &hcpclient.MetricsConfig{
+			Endpoint: &url.URL{
+				Scheme: "http",
+				Host:   "localhost",
+				Path:   "/test",
+			},
+		},
+		RefreshConfig: &hcpclient.RefreshConfig{},
+	}, nil)
+	mockClient.EXPECT().GetObservabilitySecret(mock.Anything).Return("xxx", "yyy", nil)
+	mgr.Register(TelemetryStateController(mockClientFn))
+	mgr.SetRaftLeader(true)
+	go mgr.Run(suite.ctx)
+
+	linkData := &pbhcp.Link{
+		ClientId:     "abc",
+		ClientSecret: "abc",
+		ResourceId:   "abc",
+	}
+
+	link := rtest.Resource(pbhcp.LinkType, "global").
+		WithData(suite.T(), linkData).
+		WithStatus(link.StatusKey, &pbresource.Status{Conditions: []*pbresource.Condition{link.ConditionLinked(linkData.ResourceId)}}).
+		Write(suite.T(), suite.client)
+
+	suite.T().Cleanup(suite.deleteResourceFunc(link.Id))
+
+	tsRes := suite.client.WaitForResourceExists(suite.T(), &pbresource.ID{Name: "global", Type: pbhcp.TelemetryStateType})
+	decodedState, err := resource.Decode[*pbhcp.TelemetryState](tsRes)
+	require.NoError(suite.T(), err)
+	require.Equal(suite.T(), linkData.ResourceId, decodedState.GetData().ResourceId)
+	require.Equal(suite.T(), "xxx", decodedState.GetData().ClientId)
+	require.Equal(suite.T(), "http://localhost/test", decodedState.GetData().Metrics.Endpoint)
+
+	suite.client.MustDelete(suite.T(), link.Id)
+	suite.client.WaitForDeletion(suite.T(), tsRes.Id)
+}
+
+func (suite *controllerSuite) TestController_LinkingDisabled() {
+	// Run the controller manager
+	mgr := controller.NewManager(suite.client, suite.rt.Logger)
+	_, mockClientFn := mockHcpClientFn(suite.T())
+	mgr.Register(TelemetryStateController(mockClientFn))
+	mgr.SetRaftLeader(true)
+	go mgr.Run(suite.ctx)
+
+	linkData := &pbhcp.Link{
+		ClientId:     "abc",
+		ClientSecret: "abc",
+		ResourceId:   "abc",
+	}
+
+	linkRes := rtest.Resource(pbhcp.LinkType, "global").
+		WithData(suite.T(), linkData).
+		WithStatus(link.StatusKey, &pbresource.Status{Conditions: []*pbresource.Condition{link.ConditionDisabled}}).
+		Write(suite.T(), suite.client)
+
+	suite.client.WaitForReconciliation(suite.T(), linkRes.Id, link.StatusKey)
+	suite.client.RequireResourceNotFound(suite.T(), &pbresource.ID{Name: "global", Type: pbhcp.TelemetryStateType})
```

Comment:
I think this would also work and would reduce the possibility of the test flaking:
```suggestion
	suite.client.WaitForDeletion(suite.T(), &pbresource.ID{Name: "global", Type: pbhcp.TelemetryStateType})
```

My understanding is there's some time in between when the link reconciliation happens and when the TelemetryState gets deleted so as written, I think the test could fail sometimes.